File: Questão 3/NodeJS/fibonacci.js
const fs = require('fs');

const memo = new Array(100).fill(0);

function fibonacci(n)
{
    if (n <= 1)
        return n;

    let prev = 0, atual = 1, prox;

    if (memo[n] !== 0) return memo[n];

    for (let i = 2; i <= n; i++)
    {
        if (memo[i] !== 0)
        {
            prev = atual;
            atual = memo[i];
            continue;
        }

        prox = prev + atual;
        prev = atual;
        atual = prox;

        memo[i] = atual;
    }

    memo[n] = atual;

    return atual;
}


async function main()
{
    const tempoExecucao = [];

    for (let i = 1; i <= 100; i++)
    {
        const inicio = process.hrtime.bigint();

        fibonacci(90);

        const fim = process.hrtime.bigint();

        tempoExecucao.push(Number(fim - inicio) / 1e9);
        fs.appendFileSync('fibonacci.txt', tempoExecucao[tempoExecucao.length - 1].toFixed(10) + "\n");
    }

    tempoExecucao.sort((a, b) => a - b);

    let media = 0;
    for (let i = 5; i < 95; i++)
        media += tempoExecucao[i];
    media /= 90;


    const log = 'Fibonacci iterativo de 90' + fibonacci(90) + '\n\n' +
        'Pior tempo de execução: ' + Math.max(...tempoExecucao).toFixed(10) + 's\n' +
        'Melhor tempo de execução: ' + Math.min(...tempoExecucao).toFixed(10) + 's\n' +
        'Tempo médio de execução (90 execuções): ' + media.toFixed(10) + 's\n';

    console.log(log);
}

main();

File: Questão 3/Python/fibonacci.py
import time
import numpy as np

memo = [0] * 100

def fibonacci(n):
    if n <= 1:
        return n

    prev, atual = 0, 1

    if memo[n] != 0:
        return memo[n]

    for i in range(2, n + 1):
        if memo[i] != 0:
            prev = atual
            atual = memo[i]
            continue

        prox = prev + atual
        prev = atual
        atual = prox

        memo[i] = atual

    memo[n] = atual

    return atual

def main():
    tempo_execucao = []

    for _ in range(1, 101):
        inicio = time.perf_counter_ns()

        fibonacci(90)

        fim = time.perf_counter_ns()

        tempo_execucao.append((fim - inicio) / 1e9)
        with open('fibonacci.txt', 'a') as f:
            f.write(f"{tempo_execucao[-1]:.10f}\n")

    tempo_execucao.sort()

    media = np.mean(tempo_execucao[5:95])

    log = (
        f'Fibonacci iterativo de 90: {fibonacci(90)}\n\n'
        f'Pior tempo de execução: {max(tempo_execucao):.10f}s\n'
        f'Melhor tempo de execução: {min(tempo_execucao):.10f}s\n'
        f'Tempo médio de execução (90 execuções): {media:.10f}s\n'
    )

    print(log)

if __name__ == "__main__":
    main()
